<h1 style="text-align: center;">PROJECT NEO AISSYR</h1>

<img 
    style="display: block; 
           margin-left: auto;
           margin-right: auto;
           width: 40%;"
    src="/Users/charley.dlebarbier/Documents/Dev/VSCode/Projects/projet_NeoAIssyr/extra/cover.jpg" 
    alt="æventyr illustration">
</img>

<h3 style="margin-left: 80px; font-style: bold;">ABSTRACT : FR - EN</h3>
<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
Neo AISSYR est une <b>web application</b> permettant aux chercheurs, spécialistes ou curieux de l'empire Assyrien, de transcrire les tablettes en cunéiforme de la période dite Néo-Assyrienne s'étendant entre 900 et 600 ans avant notre ère.

<p style="text-align: center;">-*-</p>
<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
Neo AISSYR is a <b>web application</b> enabling researchers, specialists or those curious about the Assyrian Empire, to transcribe cuneiform tablets from the so-called Neo-Assyrian period stretching from 900 to 600 BC.

</br>

<p style="text-align: center;">-*-</p>
<h4 style="margin-left: 80px; font-style: bold;">Source dataset</h4>
<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
<a href="https://github.com/CompVis/cuneiform-sign-detection-dataset">Github Dataset</a></p>
<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
Introduit dans : <em>Dencker, T., Klinkisch, P., Maul, S. M., and Ommer, B. (2020): Deep Learning of Cuneiform Sign Detection with Weak Supervision using Transliteration Alignment, PLOS ONE, 15:12, pp. 1–21</em> <a href="https://doi.org/10.1371/journal.pone.0243039">article link</a></p>

---


<h4 style="text-align: center;">NOTEBOOK PARTIE 1 - PART 1</h4>
<p style="text-align: center; font-style: italic">~~ ANALYSE EXPLORATOIRE DU DATASET - DATASET EDA ~~</p>
</br>

<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
Ce projet s'appuie sur le dataset présenté dans l'article <em>Dencker, T., Klinkisch, P., Maul, S. M., and Ommer, B. (2020): Deep Learning of Cuneiform Sign Detection with Weak Supervision using Transliteration Alignment, PLOS ONE, 15:12, pp. 1–21</em> est disponible sur <a href="https://github.com/CompVis/cuneiform-sign-detection-dataset">Github</a>. L'ensemble des données repose sur plus de 1800 images de tablettes d'argile de l'époque néo-assyrienne de l'écriture cunéiforme (900 - 600 ans AV. N. È.). Le dataset, fournit les Bounding Box de 8109 signes labellisés en 189 classes, réparties sur 81 tablettes.</p>

<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
Le présent notebook permet de télécharger les fichiers en vue d'initier, dans un second temps, l'analyse exploratoire du dataset.
</p>

</br>

---
</br>

In [57]:
import git
import numpy as np
import os
import pandas as pd
import re
import shutil

## Récupérer le dataset depuis github

In [2]:
GIT_REPO = "https://github.com/CompVis/cuneiform-sign-detection-dataset.git"
FOLDER_PATH = '../data'

In [3]:
git.Repo.clone_from(GIT_REPO, 
                    FOLDER_PATH,
                    branch='master')

<git.repo.base.Repo '/Users/charley.dlebarbier/Documents/Dev/VSCode/Projects/projet_NeoAIssyr/data/.git'>

## Suppression des Fichiers et Dossiers inutiles

<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
Une fois le répertoire téléchargé, il est recommandé d'analyser sa structure en vue de supprimer d'éventuels éléments inutiles.  
</p>

In [ ]:
!rptree -d '../data'

<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
Suite à ce premier examen, le dossier .git n'étant pas nécessaire, il est recommandé de le supprimer. 
<p>

In [18]:
HIDDEN_GIT_PATH = '../data/.git'

try:
    shutil.rmtree(HIDDEN_GIT_PATH)
except Exception as e:
    print(f"Error : {e}")

<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
À la suite de la suppresion, l'examen va se porter cette fois-ci sur les fichiers
</p>

In [20]:
!rptree '../data'

../data/
│
├── annotations/
│   ├── bbox_annotations_saa05.csv
│   ├── bbox_annotations_saa06.csv
│   ├── bbox_annotations_saa09.csv
│   ├── bbox_annotations_test_full.csv
│   ├── bbox_annotations_train_A.csv
│   ├── bbox_annotations_train_B.csv
│   ├── bbox_annotations_train_C.csv
│   ├── bbox_annotations_train_D.csv
│   ├── bbox_annotations_train_E.csv
│   ├── bbox_annotations_train_F.csv
│   ├── bbox_annotations_train_full.csv
│   ├── line_annotations_test.csv
│   └── line_annotations_train.csv
│
├── images/
│   │
│   └── test/
│       └── tablet_segment_annotated.png
│
│
├── segments/
│   ├── tablet_segments_saa01.csv
│   ├── tablet_segments_saa05.csv
│   ├── tablet_segments_saa06.csv
│   ├── tablet_segments_saa08.csv
│   ├── tablet_segments_saa09.csv
│   ├── tablet_segments_saa10.csv
│   ├── tablet_segments_saa13.csv
│   ├── tablet_segments_saa16.csv
│   ├── tablet_segments_test.csv
│   └── tablet_segments_train.csv
│
├── transliterations/
│   ├── transliterations_saa01.csv
│   ├─

<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
Ce deuxième examen montre que certains éléments, tels que les fichiers README, config, etc ou encore le dossier test présent dans "images" ne sont également pas utiles.  
Il seront ainsi supprimer pour de garder un environnement de travail propice à la suite.
<p>

In [23]:
DELETE_FILES_PATH = ['../data/_config.yml', 
                     '../data/Cuneiform_Sign_Detection_Dataset_Introduction.ipynb',
                     '../data/README.md']

try:
    ## - delete 'test' folder in 'images'
    shutil.rmtree('../data/images/test/')

    ## - delete _config.yml, notebook and readme
    for file in DELETE_FILES_PATH:
        os.remove(file)

except Exception as e:
    print(f"Error : {e}")

<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
Ce deuxième examen permet également de mettre en lumière trois sous-dossiers, dans lesquels les fichiers CSV sont répartis.
<p>

<p style="text-align: justify; margin-left: 80px; margin-right: 80px;"><b>- Annotations : </b>contient les fichiers csv des bbox des signes pour les 81 tablettes, et les fichiers csv pour délimiter les lignes (1er signe au dernier signe)</p>  

<p style="text-align: justify; margin-left: 80px; margin-right: 80px;"><b>- Segments : </b>contient les fichiers csv pour recadrer les tablettes sur le segment étudié</p>  

<p style="text-align: justify; margin-left: 80px; margin-right: 80px;"><b>- Transliteration : </b>contient les fichiers csv pour faire le lien avec les transliterations de Oracc</p>

## Dossier Annotations

<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
À présent examinons le fichier bbox_annotations_train_full du dossier annotations.
</p>
<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
<em>Pour réduire l'EDA, le choix a été fait de se concentrer sur les fichiers qui seront utilisés par la suite</em>
</p>

In [25]:
df_bb_ann = pd.read_csv('../data/annotations/bbox_annotations_train_full.csv')

In [29]:
df_bb_ann.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4708 entries, 0 to 4707
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   segm_idx       4708 non-null   int64 
 1   tablet_CDLI    4708 non-null   object
 2   view_desc      4361 non-null   object
 3   collection     4708 non-null   object
 4   mzl_label      4708 non-null   int64 
 5   train_label    4708 non-null   int64 
 6   bbox           4708 non-null   object
 7   relative_bbox  4708 non-null   object
dtypes: int64(3), object(5)
memory usage: 294.4+ KB


In [26]:
df_bb_ann.head()

segm_idx tablet_CDLI view_desc collection  mzl_label  train_label  \
0      2061     P334894       Obv      saa05        839           11   
1      2061     P334894       Obv      saa05        110            4   
2      2061     P334894       Obv      saa05         18          210   
3      2061     P334894       Obv      saa05         24           58   
4      2061     P334894       Obv      saa05        812           82   

                      bbox         relative_bbox  
0   [805, 1979, 871, 2055]   [165, 60, 231, 136]  
1   [860, 1988, 981, 2049]   [220, 69, 341, 130]  
2   [811, 2056, 948, 2133]  [171, 137, 308, 214]  
3  [927, 2059, 1046, 2131]  [287, 140, 406, 212]  
4   [804, 2136, 908, 2220]  [164, 217, 268, 301]

<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
Le fichier CSV contient 4708 entrées et est structuré en 8 colonnes, fournissant les informations suivantes :
</p>

<p style="text-align: justify; margin-left: 80px; margin-right: 80px;"><b>- segm_idx :</b> l'index du segment déterminant le côté de la tablette étudiée (fichiers CSV du dossier segments)</p>
<p style="text-align: justify; margin-left: 80px; margin-right: 80px;"><b>- tablet_CDLI :</b> numéro de référence de la tablette sur le site de la CDLI</p>
<p style="text-align: justify; margin-left: 80px; margin-right: 80px;"><b>- view_desc :</b> côté de la tablette étudiée -> avant : Obs ; arrière : Rev</p>
<p style="text-align: justify; margin-left: 80px; margin-right: 80px;"><b>- collection :</b> code du corpus SAAo (State Archives of Assyria online)</p>
<p style="text-align: justify; margin-left: 80px; margin-right: 80px;"><b>- mzl_label :</b> code du signe selon la classification <em>"Mesopotamisches Zeichenlexikon"</em> de l'assyriologue Rykle Borge</p>
<p style="text-align: justify; margin-left: 80px; margin-right: 80px;"><b>- train_label :</b> label pour l'apprentissage du modèle d'IA</p>
<p style="text-align: justify; margin-left: 80px; margin-right: 80px;"><b>- bbox :</b> coordonnées x_min, y_min, x_max, y_max de la bbox du signe, selon la taille réelle de l'image</p>
<p style="text-align: justify; margin-left: 80px; margin-right: 80px;"><b>- relative_bbox :</b> coordonnées x_min, y_min, x_max, y_max de la bbox du signe, selon le recadrage de la tablette pour ne prendre en compte que la view_desc</p>


In [64]:
list_full_tablet = np.array(np.unique(df_bb_ann['tablet_CDLI']))

print(f"Tablet CDLI (Unique Reference) : {len(list_full_tablet)}")
print(f"Tablet CDLI List : {list_full_tablet}")

Tablet CDLI (Unique Reference) : 47
Tablet CDLI List : ['K01057' 'P239304' 'P313634' 'P314346' 'P314355' 'P333952' 'P334311'
 'P334357' 'P334828' 'P334831' 'P334836' 'P334839' 'P334842' 'P334848'
 'P334863' 'P334865' 'P334892' 'P334894' 'P334896' 'P334921' 'P335598'
 'P335605' 'P335608' 'P336178' 'P336608' 'P336614' 'P336634' 'P336635'
 'P336643' 'P336651' 'P336655' 'P336660' 'P336663b' 'P336670' 'P336678'
 'P336716' 'P337153' 'P337163' 'P393668' 'P404881' 'VAT08803Rs'
 'VAT08803Vs' 'VAT10601' 'VAT10657' 'VAT11100Rs' 'VAT13604Rs' 'VAT13604Vs']


<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
Dans ce dataframe, nous avons 47 références de tablettes qui sont uniques. Étant donné qu'il comprend l'ensemble des tablettes, il est bon de vérifier qu'il est bien composé de toutes celles présentes dans les autres fichiers CSV 'bbox_annotations_train_*'. Ce contrôle garantira que ce dataframe, seul, soit suffisant pour l'apprentissage du modèle ou pour le fusionner avec celui de test, si le souhait de faire un nouveau split est envisagé. En cas de tablettes manquantes, elles pourront alors être ajouter.
</p>
<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
La présence de toutes les références dans le fichier 'full' se fera en contrôlant la présence des lignes des fichiers dans celui du 'full'
</p>

In [80]:
FOLDER_PATH = '../data/annotations/'
FILES_SEARCH = re.compile(r'^bbox_annotations_train_[A-F]')

bbox_csv_list = sorted([os.path.join(FOLDER_PATH, csv) for csv in os.listdir(FOLDER_PATH) \
                 if os.path.isfile(os.path.join(FOLDER_PATH, csv)) and FILES_SEARCH.match(csv)])

In [81]:
for csv in bbox_csv_list:
    df = pd.read_csv(csv)

    merged_df = pd.merge(df, df_bb_ann, how='inner', indicator=True)
    mismatch_rows = merged_df[merged_df['_merge'] != 'both']

    if mismatch_rows.empty:
        print(f"All lines from {csv} are in bbox_annotations_train_full.")
    else:
        print(f"Lines from {csv} not in bbox_annotations_train_full :")
        print(mismatch_rows)

All lines from ../data/annotations/bbox_annotations_train_A.csv are in bbox_annotations_train_full.
All lines from ../data/annotations/bbox_annotations_train_B.csv are in bbox_annotations_train_full.
All lines from ../data/annotations/bbox_annotations_train_C.csv are in bbox_annotations_train_full.
All lines from ../data/annotations/bbox_annotations_train_D.csv are in bbox_annotations_train_full.
All lines from ../data/annotations/bbox_annotations_train_E.csv are in bbox_annotations_train_full.
All lines from ../data/annotations/bbox_annotations_train_F.csv are in bbox_annotations_train_full.


<p style="text-align: justify; margin-left: 80px; margin-right: 80px;">
La vérification attestant la présence de toutes les lignes dans le fichier csv 'Full', il est possible de continuer l'analyse en se focalisant que sur ce dernier.</p>

In [82]:
df_bb_ann.isnull().sum()

segm_idx           0
tablet_CDLI        0
view_desc        347
collection         0
mzl_label          0
train_label        0
bbox               0
relative_bbox      0
dtype: int64